outline.md

```diff
@@ -1,10 +1,55 @@
+https://stackoverflow.com/questions/48311645/installing-tensorflow-using-pip-within-anaconda-fails
+
+### Example: Composing Environments for Polynomial Solves
+
+A solution Twitter once had was a partitioning-based constraint solver which kept track of all available versions of each target, using a union-find as well as locally merging constraint requirements. This is described because it introduces the concept of *merging* which is adapted here in [filesystem operations](#filesystem-operations). It was never used in production.
+
+[pants](https://github.com/pantsbuild/pants) lets you organize and publish your python code into multiple independent wheels published to PyPI. My technique created a pip lockfile for every root python project, then propagated matching requirement constraints back up to parent libraries, all the way up to the dependency roots. This amortized the cost of having to actually ping PyPI for new requirements to occur literally only when a single conflict was found. More importantly, this process separated for the first time the user and maintainer UX, where *users* have no responsibilities and don't wait very long, and *maintainers* are responsible for using *domain knowledge* to make a requirement update go as smoothly as possible (with as few one-off hacks as possible):
+- *user*: pulls master. makes changes to source code.
+    - if no changes to requirements, then the install is retrieved from cache (instant).
+    - if requirements change on some maybe-root target, a lockfile is generated for that target, and then that lockfile is used to resolve and populate the cache. This lockfile can then be propagated back *down* to all roots.
+- *maintainer*: "ok, someone wants to update this dependency, i'll figure out how to make that work."
+    - Applies a diff that bumps tensorflow to version 2 (this would be done by a member on the machine learning team).
+    - Re-run pip (in parallel) over every root target again, and merge those constraints up to recursive parents when possible (does not hit the internet).
+    - For each case where merging constraints produces an empty `VersionRange`, that implies a contradiction has occurred.
+    - Scan pypi for all nearby applicable (according to platform) versions of the package, and attempts to find a neighborhood around the previous `VersionRange` value of that constraints which *does* confirm to the requirements.
+    - For all cases where this automatic local search fails, create a pants target which points back to the previous requirement version, before updating it, and point all such cases to the old requirement version. These are considered TODOs.
+
+So after spending some *pants developer* time, we were able to make the user UX **strictly better than pip**, and we developed a runbook which I hope means users with domain experience (aka *maintainers*) no longer have to do anything in their head except focus purely on the spots that weren't fixed automatically.
```

Comment:
Note that the model doc isn't about UX -- it's an important topic but out of scope.

Reply:
I'm trying to distinguish how users, package developers, and maintainers interact. The "goal" of this resolve process can be considered as optimizing the UX for users. This allows us to compare the tradeoffs of separate solutions e.g. the gradual pip resolve with no external code vs the monolithic concretization with clingo.